File: 234.palindrome-linked-list.py
#
# @lc app=leetcode id=234 lang=python3
#
# [234] Palindrome Linked List
#
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def isPalindrome(self, head: ListNode) -> bool:        
        if head == None or head.next == None: return True

        slow, fast = head, head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        
        rev_sec_half = self.reverse_list(slow.next)
        slow.next = None
        
        while rev_sec_half:
            if head.val != rev_sec_half.val:
                return False
            head = head.next
            rev_sec_half = rev_sec_half.next
        
        return True
            
    def reverse_list(self, head):

        pre = None
        cur = head
        while cur:
            nxt = cur.next
            cur.next = pre
            pre = cur
            cur = nxt

        return pre

File: 234.palindrome-linked-list.java
/*
 * @lc app=leetcode id=234 lang=java
 *
 * [234] Palindrome Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isPalindrome(ListNode head) {
        
        if (head == null) {
            return true;
        }
        
        // only one node
        if (head != null && head.next == null) {
            return true;
        }
        
        ListNode slow = head;
        ListNode fast = head.next;
        
        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }
        
        ListNode secHead = slow.next;
        slow.next = null;
        
        secHead = reverseList(secHead);
        
        while (head != null && secHead != null) {
            if (head.val != secHead.val) {
                return false;
            }
            head = head.next;
            secHead = secHead.next;
        }
        
        return true;
    }
    
    public ListNode reverseList(ListNode head) {
		ListNode pre = null;
		ListNode cur = head;

		while (cur != null) {
			ListNode nxt = cur.next;
			cur.next = pre;
			pre = cur;
			cur = nxt;
		}

		return pre;
	}
}
// @lc code=end

